Mathlib/Combinatorics/SimpleGraph/Clique.lean

```diff
@@ -114,6 +111,72 @@ theorem isClique_bot_iff : (⊥ : SimpleGraph α).IsClique s ↔ (s : Set α).Su
 alias ⟨IsClique.subsingleton, _⟩ := isClique_bot_iff
 #align simple_graph.is_clique.subsingleton SimpleGraph.IsClique.subsingleton
 
+protected theorem IsClique.map (h : G.IsClique s) {f : α ↪ β} : (G.map f).IsClique (f '' s) := by
+  rintro _ ⟨a, ha, rfl⟩ _ ⟨b, hb, rfl⟩ hab
+  exact ⟨a, b, h ha hb <| ne_of_apply_ne _ hab, rfl, rfl⟩
+#align simple_graph.is_clique.map SimpleGraph.IsClique.map
+
+theorem isClique_map_iff_of_nontrivial {f : α ↪ β} {t : Set β} (ht : t.Nontrivial) :
+    (G.map f).IsClique t ↔ ∃ (s : Set α), G.IsClique s ∧ f '' s = t := by
+  refine ⟨fun h ↦ ⟨f ⁻¹' t, ?_, ?_⟩, by rintro ⟨x, hs, rfl⟩; exact hs.map⟩
+  · rintro x (hx : f x ∈ t) y (hy : f y ∈ t) hne
+    obtain ⟨u,v, huv, hux, hvy⟩ := h hx hy (by simpa)
+    rw [EmbeddingLike.apply_eq_iff_eq] at hux hvy
+    rwa [← hux, ← hvy]
+  rw [Set.image_preimage_eq_iff]
+  intro x hxt
+  obtain ⟨y,hyt, hyne⟩ := ht.exists_ne x
+  obtain ⟨u,v, -, rfl, rfl⟩ := h hyt hxt hyne
+  exact Set.mem_range_self _
+
+theorem isClique_map_iff {f : α ↪ β} {t : Set β} :
+    (G.map f).IsClique t ↔ t.Subsingleton ∨ ∃ (s : Set α), G.IsClique s ∧ f '' s = t := by
+  obtain (ht | ht) := t.subsingleton_or_nontrivial
+  · simp [isClique_subsingleton _ ht, ht]
+  simp [isClique_map_iff_of_nontrivial ht, ht.not_subsingleton]
+
+@[simp] theorem isClique_map_image_iff {f : α ↪ β} :
+    (G.map f).IsClique (f '' s) ↔ G.IsClique s := by
+  rw [isClique_map_iff, f.injective.subsingleton_image_iff]
+  obtain (hs | hs) := s.subsingleton_or_nontrivial
+  · simp [hs, isClique_subsingleton G hs]
+  simp [or_iff_right hs.not_subsingleton, Set.image_eq_image f.injective]
+
+theorem isClique_of_card_le_one (G : SimpleGraph α) {t : Finset α} (ht : t.card ≤ 1) :
+    G.IsClique t := by
+  apply isClique_subsingleton
+  rw [Finset.card_le_one_iff_subsingleton_coe] at ht
+  exact (Set.subsingleton_coe _).1 ht
+
+section DecidableEq
+
+variable [DecidableEq β] {f : α ↪ β} {t : Finset β}
+
+theorem isClique_map_finset_iff_of_nontrivial (ht : t.Nontrivial) :
+    (G.map f).IsClique t ↔ ∃ (s : Finset α), G.IsClique s ∧ s.map f = t := by
+  constructor
+  · rw [isClique_map_iff_of_nontrivial (by simpa)]
+    rintro ⟨s, hs, hst⟩
+    obtain ⟨s, rfl⟩ := Set.Finite.exists_finset_coe <|
+      (show s.Finite from Set.Finite.of_finite_image (f := f) (by simp [hst]) (f.injective.injOn _))
+    exact ⟨s,hs, Finset.coe_inj.1 (by simpa)⟩
+  rintro ⟨s, hs, rfl⟩
+  simpa using hs.map (f := f)
+
+theorem isClique_map_finset_iff :
+    (G.map f).IsClique t ↔ t.card ≤ 1 ∨ ∃ (s : Finset α), G.IsClique s ∧ s.map f = t := by
+  obtain (ht | ht) := le_or_lt t.card 1
+  · simp [ht, isClique_of_card_le_one]
+  rw [isClique_map_finset_iff_of_nontrivial, ← not_lt]
+  · simp [ht, Finset.map_eq_image]
+  exact Finset.one_lt_card_iff_nontrivial.mp ht
+
+protected theorem IsClique.finset_map {f : α ↪ β} {s : Finset α} (h : G.IsClique s) :
```

Comment:
```suggestion
protected theorem IsClique.map {f : α ↪ β} {s : Finset α} (h : G.IsClique s) :
```
and rename the existing `IsClique.map` to `IsClique.image`, maybe?

Reply:
I think this is prioritizing `map` to be first and foremost something that refers to finsets over something that refers to simple graphs, which doesn't seem like the correct choice in the `SimpleGraph` namespace/API. (If we were in the `Finset` namespace, I'd agree.)

I do see both sides of this, though - it probably should be discussed further rather than changing it here.

Reply:
Then for now at least
```suggestion
protected theorem IsClique.finsetMap {f : α ↪ β} {s : Finset α} (h : G.IsClique s) :
```

Reply:
> I think this is prioritizing `map` to be first and foremost something that refers to finsets over something that refers to simple graphs, which doesn't seem like the correct choice in the `SimpleGraph` namespace/API

Thinking about again, does it not exactly make sense here in that cliques are (almost) always finsets?

Reply:
I wouldn't say they are.

Reply:
At least - their definition has nothing inherently to do with finiteness.

Reply:
Again, I think this discussion can happen elsewhere.